src/rstrtmgr.cpp

```diff
@@ -0,0 +1,88 @@
+/* ****************************************************************************
+*
+* Copyright (c) Microsoft Corporation. All rights reserved.
+*
+*
+* This file is part of Microsoft R Host.
+*
+* Microsoft R Host is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Microsoft R Host is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Microsoft R Host.  If not, see <http://www.gnu.org/licenses/>.
+*
+* ***************************************************************************/
+
+#include "rstrtmgr.h"
+#include "Rapi.h"
+#include "util.h"
+#include <RestartManager.h>
+
+namespace rhost {
+    namespace util {
+        file_lock_state lock_state_by_file(const std::vector<std::string>& paths) {
+#ifdef _WIN32
+            std::vector<std::wstring> wpaths(paths.size());
+            std::transform(paths.begin(), paths.end(), wpaths.begin(), [](const std::string& path) { return to_unicode(path); });
+            std::vector<LPCWSTR> wfilepaths(wpaths.size());
+            std::transform(wpaths.begin(), wpaths.end(), wfilepaths.begin(), [](std::wstring& wpath) { return wpath.data(); });
+
+            std::unique_ptr<WCHAR> session_uuid_str(new WCHAR[128]);
+            DWORD session_handle = 0;
+            SCOPE_WARDEN(rstrtmgr_end, {
+                if (session_handle) {
+                    RmEndSession(session_handle);
+                }
+            });
+
+            DWORD error = RmStartSession(&session_handle, 0, session_uuid_str.get());
+            if (error != ERROR_SUCCESS) {
+                return file_lock_state::unlocked;
+            }
+
+            error = RmRegisterResources(session_handle, static_cast<UINT>(wfilepaths.size()), wfilepaths.data(), 0, nullptr, 0, nullptr);
+            if (error != ERROR_SUCCESS) {
+                return file_lock_state::unlocked;
+            }
+
+            UINT process_info_needed = 0;
+            UINT process_info_count = 0;
+            DWORD reboot_reason = RmRebootReasonNone;
+            error = RmGetList(session_handle, &process_info_needed, &process_info_count, nullptr, &reboot_reason);
+            if (error != ERROR_SUCCESS && error != ERROR_MORE_DATA) {
+                return file_lock_state::unlocked;
+            }
+
+            std::vector<DWORD> process_ids;
```

Comment:
Looks like this isn't used inside the loop that follows - so it can be declared after the loop, and get the size directly when constructed (instead of potentially being resized several times in the loop).